### PR TITLE
plat: imx8mp: Add the src handler for m7 core boot support

### PR DESCRIPTION
**Pull request by Qingwu-Li**
Add the SRC SiP handler for M7 boot support on i.MX8MP SoC

Signed-off-by: LI Qingwu <Qing-wu.Li@leica-geosystems.com.cn>

### CODE REVIEW

File: plat/imx/common/imx_sip_svc.c [hunk 1]
```diff
@@ -34,6 +34,11 @@ static uintptr_t imx_sip_handler(unsigned int smc_fid,
 		SMC_RET1(handle, imx_soc_info_handler(smc_fid, x1, x2, x3));
 		break;
 #endif
+#if defined(PLAT_imx8mm) || defined(PLAT_imx8mn) || defined(PLAT_imx8mp)
+	case IMX_SIP_SRC:
+		SMC_RET1(handle, imx_src_handler(smc_fid, x1, x2, x3, handle));
+		break;
+#endif
```
Comment (zandrey) [inline plat/imx/common/imx_sip_svc.c:41]: NAK, this is already done below, see line 57 below.

Extend the conditional compilation with `PLAT_imx8mn` and `PLAT_imx8mp` instead.

File: plat/imx/common/include/imx_sip_svc.h [hunk 2]
```diff
@@ -18,6 +18,9 @@
 #define IMX_SIP_BUILDINFO_GET_COMMITHASH	0x00
 
 #define IMX_SIP_SRC			0xC2000005
+#define IMX_SIP_SRC_M4_START		0x00
+#define IMX_SIP_SRC_M4_STARTED		0x01
+#define IMX_SIP_SRC_M4_STOP		0x02
```
Comment (zandrey) [inline plat/imx/common/include/imx_sip_svc.h:23]: Is this aligned with _U-Boot_? From the commit message, it is not clear.

File: plat/imx/imx8m/gpc_common.c [hunk 3]
```diff
@@ -250,3 +254,70 @@ void imx_clear_rbc_count(void)
 	mmio_clrbits_32(IMX_GPC_BASE + SLPCR, SLPCR_RBC_EN |
 		(0x3f << SLPCR_RBC_COUNT_SHIFT));
 }
+#pragma weak imx_src_handler
+/* imx8mq/imx8mm need to verrride below function */
+int imx_src_handler(uint32_t smc_fid, u_register_t x1, u_register_t x2,
```
Comment (zandrey) [inline plat/imx/imx8m/gpc_common.c:259]: NAK, this is already present in `plat/imx/common/imx_sip_handler.c`

File: plat/imx/imx8m/imx8mp/imx8mp_bl31_setup.c [hunk 4]
```diff
@@ -207,6 +207,10 @@ void bl31_platform_setup(void)
 	plat_gic_init();
 
 	imx_gpc_init();
+
+	/* Enable and reset M7 */
+	mmio_setbits_32(IMX_SRC_BASE + 0xc, SRC_SCR_M4_ENABLE_MASK);
+	mmio_clrbits_32(IMX_SRC_BASE + 0xc, SRC_SCR_M4C_NON_SCLR_RST_MASK);
```
Comment (zandrey) [inline plat/imx/imx8m/imx8mp/imx8mp_bl31_setup.c:213]: Does `PLAT_imx8mm` require the same?

**Review by zandrey — CHANGES_REQUESTED** [review]
Is this a cherry-pick of the downstream commit? Are we sure that everything present here is required?

File: plat/imx/common/include/imx_sip_svc.h (hunk 2 above)
Comment (zandrey) [reply plat/imx/common/include/imx_sip_svc.h:23]: Found the info myself: this is already present in the _U-Boot_ tree in `include/imx_sip.h`. This was introduced with commit [`de274663f542 ("imx: add sip function")`](https://github.com/hexagon-geo-surv/u-boot/commit/de274663f5429fe10c48a40f9fbf0c8aa553905d).

`IMX_SIP_SRC_M4_STOP` came with commit [`c383379f25e2 ("imx: imx9: Support booting m33 from Acore")`](https://github.com/hexagon-geo-surv/u-boot/commit/c383379f25e2491d9502f367324744cc580c0eb2)

File: plat/imx/common/imx_sip_svc.c [hunk 5]
```diff
@@ -33,6 +33,9 @@ static uintptr_t imx_sip_handler(unsigned int smc_fid,
 	case IMX_SIP_GET_SOC_INFO:
 		SMC_RET1(handle, imx_soc_info_handler(smc_fid, x1, x2, x3));
 		break;
+	case IMX_SIP_SRC:
+		SMC_RET1(handle, imx_src_handler(smc_fid, x1, x2, x3, handle));
+		break;
```
Comment (zandrey) [inline plat/imx/common/imx_sip_svc.c:38]: Please drop this block, `PLAT_imx8mq` is already defined below.

File: plat/imx/common/imx_sip_svc.c [hunk 6]
```diff
@@ -48,7 +51,7 @@ static uintptr_t imx_sip_handler(unsigned int smc_fid,
 	case IMX_SIP_MISC_SET_TEMP:
 		SMC_RET1(handle, imx_misc_set_temp_handler(smc_fid, x1, x2, x3, x4));
 #endif
-#if defined(PLAT_imx8mm) || defined(PLAT_imx8mq)
+#if defined(PLAT_imx8mm) || defined(PLAT_imx8mn) || defined(PLAT_imx8mp)
```
Comment (zandrey) [inline plat/imx/common/imx_sip_svc.c:54]: Do not drop `PLAT_imx8mq`, rather add other platforms here

File: plat/imx/imx8m/imx8mp/imx8mp_bl31_setup.c (hunk 4 above)
Comment (Qingwu-Li) [reply plat/imx/imx8m/imx8mp/imx8mp_bl31_setup.c]: it has similar here
https://github.com/Qingwu-Li/trusted-firmware-a/blob/3656720985b2bfbfdf379949d4802fd645f6418d/plat/imx/imx8m/gpc_common.c#L262

**Pull request closed without merging**